### ISSUE TITLE
Infinite loop with corecord

### ISSUE THREAD

**Issue by josefs** [idris-lang/Idris-dev#2333]
Here's some code:

~~~
corecord Point where
  constructor MkPoint
  x : Int
  y : Int
  move : Int -> Int -> Point

point : Int -> Int -> Point
point x y = MkPoint x y (\dx, dy => point (x + dx) (y + dy))
~~~

The interesting thing in this code is `move` which creates a new point and makes the record definition (co-)recursive. Trouble begins when I try to use `move` in the repl like so:

    > move (point 5 5) 3 2

Idris spins into an infinite loop. I'm surpised because there's nothing infinite about that expression. I suspect that it might be the printer causing the problem, because projecting one of the fields from a moved point works:

    > x (move (point 5 5) 3 2)
    8

**Comment by identicalsnowflake** [comment]
> Trouble begins when I try to use move in the repl

I don't even need to use `move`. Just typing `point 3 5` triggers the loop for me.

**Comment by david-christiansen** [comment]
As @identicalsnowflake points out, you don't need to applyt the projection `move` to get something infinite.

And, actually, `point` is creating an infinite structure, though only at compile time, and it's doing so lazily, which is why printing it dies. Let's walk through it, step by step, at least for a while. In the following, let "≡" mean "is precisely the same object as" --- that is, judgmental equality given by computation.

`point 4 3` ≡
`MkPoint 4 3 (\dx, dy => point (4 + dx) (3 + dy))` ≡
`MkPoint 4 3 (\dx, dy => MkPoint (4 + dx) (3 + dy) (\dx', dy' => point (4 + dx + dx') (3 + dy + dy')))` ≡
... and so forth. 

At compile time, Idris evaluates under all binders, because that's what gives the equality that we need for type checking. We want `\n : Nat => Z + n` to be *precisely the same thing* as `\n : Nat => n`. So Idris computes lazily in the type checker, using the compile-time evaluator, normalizing as little as possible to decide the equalities that it needs to. If it were going to check equalities of MkPoints, it might need to expand the function calls forever, depending how how it were built. So, as we'd expect, `point` does not pass the totality checker:
``
λΠ> :total point
Point.point is possibly not total due to recursive path:
    Point.point
``

The Idris REPL uses the same evaluator as the type system. However, REPL expressions need not pass the totality checker. Unlike Haskell, a term pretty-printer is used to render REPL output, rather than a `Show` instance. REPL output is completely normalized, because that's usually what people want when they're testing expressions. Even this would be OK, because that normalization is happening lazily. However, in your case, this normalization process does not terminate, because the pretty-printer forces its argument, building up that infinite lambda-term that I referred to.

You can define a `Show` instance that doesn't print the function, and then things do what you'd want:
````
instance Show Point where
  show (MkPoint x y f) = "(" ++ show x ++ ", " ++ show y ++ ")"

λΠ> show (point 2 3)
"(2, 3)" : String
λΠ> show (move (point 5 5) 3 2)
"(8, 7)" : String
````

I understand totally why this behavior is unexpected. Perhaps the REPL evaluator should impose some (configurable) maximum depth of pretty-printing in these cases.

**Comment by david-christiansen** [comment]
To make the kind of API you're looking for, I'd probably just use ordinary records:
```
record Point where
  constructor MkPoint
  x : Int
  y : Int

move : Int -> Int -> Point -> Point
move dx dy p = MkPoint (x p + dx) (y p + dy)

λΠ> :total move
Point.move is Total
```

**Comment by josefs** [comment]
Some form of improvement to the REPL printer would be good. Right now, playing with corecords is a bit awkward.

I understand that `point` is not total, but it is *productive*, i.e. it is a correct coinductive definition. Does Idris keep track of productivity in the same way it does totality? Is there a way I can ask Idris to check that a definition is productive?

Using a standard record certainly works in this case but I have code where it's not so straightforward.

**Comment by david-christiansen** [comment]
Productivity is checked using constructor-guardedness. This means that corecursive calls need to be directly under a constructor of the datatype in question, and `point` doesn't do that - instead, it guards with a lambda. Perhaps the productivity checker should be expanded to accept that as well.

The way to ask Idris to check for productivity is to annotate it as `total`. The totality checker will use productivity for coinductive data and structural size decrease for inductive data. For example, try:
```
corecord Str a where
  constructor Cons
  head : a
  tail : Str a

total
repeat : a -> Str a
repeat x = x `Cons` repeat x
```
and then ask Idris for the totality of `repeat`. It doesn't presently state the reason for the totality - that would be a nice feature.

**Comment by josefs** [comment]
The recursion in `point` is guarded by a constructor just as in `repeat`, which is why I thought that Idris would be able to figure out that it's total. But I realize the lambda can trip up the totality checker.

To give a bit of context; in my spare time I've been playing with encoding purely functional objects. It is very natural to use corecords for encoding objects, because they allow for encoding the `this` pointer.

**Comment by david-christiansen** [comment]
Constructor-guardedness is very highly conservative - the recursive call must be *immediately* under a constructor for it to work. Otherwise we could produce a constructor with an infinite loop immediately under it.

This is not a pleasant state of affairs, unfortunately.

Are you planning on writing proofs about these objects? Or are they a strictly run-time phenomenon? If it's the latter, then a pretty-printer update (or perhaps a limit to the number of unfoldings performed at the REPL) might do the trick.

**Comment by josefs** [comment]
No proofs for now. Just updating the printer would be super helpful for me.

**Comment by david-christiansen** [comment]
All right. Not sure I'll get to it today, but it's now on my shortlist.

**Comment by david-christiansen** [comment]
Once #2345 is merged, you'll be able to set a maximum depth for the printer using `:printerdepth N` at the REPL or in your init file. If you're going to be working with a lot of infinite structures, it's probably best to set it fairly low, like 100 or so, to avoid getting overwhelmed.

**Comment by josefs** [comment]
Thanks. The printerdepth feature is really helpful. For the record (no pun intended) I've had to set the constant really low, like less than 10, for the code I've been working with.

**Comment by david-christiansen** [comment]
Glad to hear that it works! It would perhaps be nice to have an alternative mode of the evaluator that doesn't evaluate under lambda or `Delay`, and a UI for making it take another step on a term that's been displayed. I think @chrisdone has done that for Haskell.

**Comment by chrisdone** [comment]
Yeah, in [the present package](http://hackage.haskell.org/package/present-2.2/docs/Present.html) there is [lazy printing](http://chrisdone.com/posts/presentations-update). It looks at constructor slots on demand.

**Comment by david-christiansen** [comment]
It probably wouldn't be difficult to build something like that on top of what idris-mode already supports for inspecting terms. Perhaps when I get a free weekend.
